Created IApplicationInfo and IApplicationStatusInfo for Dependency Injection

ahoefling opened the pull request:
Related to #3985 

## Summary
**This does not close the linked work item, but is required before I can submit a pull request.**

While working on changes to the Client Resource Manager I noticed circular dependency between `DotNetNuke.Libary` and `DotNetNuke.Web.Client`. The `ClientDependencySetting` utilizes reflection to workaround the circular dependency. Dynamically loading types like this is considered an anti-pattern, especially when we can use depenency injection to extract an interface.

This Pull Request extracts 2 interfaces from the Globals.cs
- `IApplicationInfo`
- `IApplicationStatusInfo`

ahoefling commented on DNN Platform/Library/Application/AssemblyStatusAttribute.cs:
```diff
@@ -26,27 +26,27 @@ public enum ReleaseMode
         /// Not asssigned
         /// </summary>
-        None,
+        None = 0,
```
I Added explicit numbers here to ensure this enum interops correctly with the new `DotNetNuke.Abstractions` enum. This will prevent any accidental breaking changes if the enums are ever re-ordered or changed. When we are converting the enums it is a simple cast as the int value matches, there is no need for a complex conversion method.

ahoefling commented on DNN Platform/Library/Application/AssemblyStatusAttribute.cs:
```diff
@@ -26,27 +26,27 @@ public enum ReleaseMode
         /// Alpha release
         /// </summary>
-        Alpha,
+        Alpha = 1,
```
See above comment - https://github.com/dnnsoftware/Dnn.Platform/pull/3988/files#r471134926

ahoefling commented on DNN Platform/Library/Application/DotNetNukeContext.cs:
```diff
@@ -45,30 +44,27 @@ public static DotNetNukeContext Current
+                if (current == null)
                 {
-                    _current = new DotNetNukeContext();
+                    current = Globals.DependencyProvider.GetRequiredService<IDnnContext>();
```
Should we deprecate this function and ask consumers to use Dependency Injection? I decided not to deprecate this function because I didn't add the remaining members to the new `IDnnContext`. I think once we add the remaining members we can deprecate this method and recommend Dependency Injection.

If reviewers are in agreement I can create a work item

ahoefling commented on DNN Platform/Library/Application/DotNetNukeContext.cs:
```diff
@@ -45,30 +44,27 @@ public static DotNetNukeContext Current
-                return this._application;
-            }
-        }
+        public Application Application { get => this.applicationInfo is Application app ? app : new Application(); }
```
If the class is instantiated without Dependency Injection for some reason such as unit tests, I updated a default `Application` instantiation. Maybe we should update this to the `DependencyProvider`

ahoefling commented on DNN Platform/Library/Common/Globals.cs:
```diff
@@ -316,28 +316,28 @@ public enum UpgradeStatus
             /// The application need update to a higher version.
             /// </summary>
-            Upgrade,
+            Upgrade = 0,
```
See above comment about enums - https://github.com/dnnsoftware/Dnn.Platform/pull/3988/files#r471134926

ahoefling commented on DNN Platform/Tests/DotNetNuke.Tests.Core/Controllers/Search/SearchControllerTests.cs:
```diff
@@ -139,6 +141,10 @@ public void SetUp()
             MockComponentProvider.ResetContainer();
 
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddTransient<IApplicationStatusInfo>(container => new DotNetNuke.Application.ApplicationStatusInfo(Mock.Of<IApplicationInfo>()));
```
Many of the unit tests depend on File Operations that are generated from the `MapApplicationPath` function. It was difficult to mock this out and it was easier to use an actual implementation of `IApplicationStatusInfo` instead of a mock.

ahoefling commented on DNN Platform/Library/Common/Globals.cs:
```diff
@@ -4061,5 +3830,15 @@ private static bool FilenameMatchesExtensions(string filename, string strExtensi
+        /// </summary>
+        private static void OnDependencyProviderChanged(object sender, PropertyChangedEventArgs eventArguments)
+        {
+            applicationStatusInfo = DependencyProvider.GetRequiredService<IApplicationStatusInfo>();
```
Creating a method to register dependencies once the provider has been configured is needed since the `DependencyProvider` uses our custom `LazyServiceProvider`. **It is critical that the `Global.cs` resolves only ONE instance of a dependency, as it can generate too many sql connections that causes DNN to stop functioning**. 

This finding has been documented - #3989

ahoefling commented on DNN Platform/DotNetNuke.Web/Common/LazyServiceProvider.cs:
```diff
@@ -5,28 +5,30 @@
         {
-            this._serviceProvider = serviceProvider;
+            this.serviceProvider = serviceProvider;
+            this.PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(serviceProvider)));
```
I added `INotifyPropertyChanged` here to allow the `Globals.cs` to listen for an event when the Service Provider is ready for us. Since the `Globals.cs` is wrapping several Dependency Injection calls as we deprecate methods it is important that it can resolve dependencies once they are available. This is critical when using `Globals.cs` at startup.

I decided on `INotifyPropertyChanged` as it is a .NET Standard type and is used in the desktop/mobile space a lot for notifying objects when something has changed. If there is a better option, I am more than happy to change this

ahoefling commented:
This is ready for review

I added single line comments around sections of the code that needed extra explanation of why I made some changes. All unit tests are passing. I downloaded the build artifacts and validated that it does install DNN correctly for a new installation.

ahoefling commented:
I'm not sure why it shows the build is still in progress. I just checked the build server and the build for this PR is showed success
- https://dev.azure.com/dotnet/DNN/_build/results?buildId=37348&view=results

bdukes commented on DNN Platform/Library/Application/DotNetNukeContext.cs:
```diff
@@ -1,41 +1,40 @@
-        /// <summary>
-        /// Initializes a new instance of the <see cref="DotNetNukeContext" /> class.
-        /// </summary>
-        protected DotNetNukeContext()
```
@mitchelsellers do you have any concern about a breaking change from removing this `protected` constructor?

ahoefling replied:
@bdukes is this a breaking change? the class uses a `public` access modifier, by removing the `protected` constructor it opens the API structure up slightly. I had to remove the `protected` constructor so the Dependency Injection container could properly resolve `DotNetNukeContext`. If we want to keep this `protected` I could create a new child class with a public constructor and then register `IDnnContext` to the child class.

bdukes replied:
The scenario that would break is when someone is inheriting from `DotNetNukeContext` and calling `base()` in its constructor.  It shouldn't be breaking to leave these two `protected` constructors and add the new `public` constructor, though.

ahoefling replied:
I just tried adding 2 default constructors
```
public DotNetNukeContext() {}
protected DotNetNukeContext() {}
```

This has a compilation error as a class can only have 1 parameterless default constructor regardless of access level. I don't think this is a breaking change because child classes will automatically call the default constructor regardless if it is public or protected. If the constructor was private, then you start getting into different behaviors.

bdukes replied:
I'm not seeing a default constructor on the new version of `DotNetNukeContext`, just `public DotNetNukeContext(IApplicationInfo applicationInfo)`, am I missing it?

ahoefling replied:
🤦‍♂️🤦‍♂️🤦‍♂️🤦‍♂️

Glad we talked through this, my mistake. You are 100% correct, there is no long a default constructor and that is **wrong**. We need to add a public default constructor to prevent a breaking change.

ahoefling replied:
I have added the protected constructor back in, but have deprecated it. Marking this conversation as resolved.

bdukes commented:
/azp run

ahoefling commented on DNN Platform/Library/Common/Globals.cs:
```diff
@@ -369,13 +364,8 @@ public static string ApplicationPath
-                return _applicationMapPath ?? (_applicationMapPath = GetCurrentDomainDirectory());
-            }
-        }
+        [Obsolete("Deprecated in 9.7.1. Scheduled for removal in v11.0.0, use DotNetNuke.Abstractions.IApplicationStatusInfo instead.")]
```
@bdukes @mitchelsellers 

I'm thinking of changing the verbiage of this to give better instructions. Do you have any thoughts on my suggested change?
```suggestion
        [Obsolete("Deprecated in 9.7.1. Using Dependency Injections resolve 'DotNetNuke.Abstractions.IApplicationStatusInfo' instead. Scheduled for removal in v11.0.0.")]
```

If this format is good, I'll update all the deprecation notices in this PR

ahoefling replied:
This has been fixed in the latest code push, resolving conversation

bdukes commented on DNN Platform/Library/Application/DotNetNukeContext.cs:
```diff
@@ -1,105 +1,100 @@
+        [Obsolete("Deprecated in DotNetNuke 9.7.1. This constructor has been replaced by parameterized public constructor which is designed to be used with Dependency Injection. Resolve the new interface 'DotNetNuke.Abstractions.IDnnContext' instead. Scheduled removal in v11.0.0.")]
+        protected DotNetNukeContext()
+            : this(Globals.DependencyProvider.GetRequiredService<IApplicationInfo>())
+        { }
```
```suggestion
        { }

        [Obsolete("Deprecated in DotNetNuke 9.7.1. This constructor has been replaced by the overload taking an IApplicationInfo, which should be resolved via Dependency Injection. Scheduled removal in v11.0.0.")]
    protected DotNetNukeContext(Application application)
        : this((IApplicationInfo)application)
    { }
```

ahoefling replied:
This makes sense to me, thanks for the help on the verbiage!

ahoefling commented:
@bdukes I believe I have addressed all of your concerns in the latest update. Let me know if there is anything else I need to do to get this PR ready to merge

mitchelsellers commented:
@bdukes We good for this to be merged for 9.7.1?  I don't think it will be a big/breaking issue.

bdukes commented:
![Hold on to your butts GIF](https://media.giphy.com/media/OCu7zWojqFA1W/giphy.gif)

valadas commented:
Amazing work!
